### PR TITLE
StateTimeline: Fix toolitp not showing when for frames with multiple fields

### LINKED ISSUE #42620: State timeline tooltip bug
**What happened**:
If there are more than 1 series, then there is no popup hint

**How to reproduce it (as minimally and precisely as possible)**:
Create state timeline with 2 or more rows

image

**Environment**:
- Grafana version: 8.3.0
- Data source type & version: postgresql + timescaledb
- OS Grafana is installed on: ubuntu 20
- User OS & Browser: windows 10 + latest chrome

### PR DESCRIPTION
**Pull request by dprokop**
Fixes https://github.com/grafana/grafana/issues/42620

Issue most probably caused by the query that returns a frame with multiple fields. In https://github.com/grafana/grafana/pull/39438 we introduced [a check against the equality of the **aligned data** and the **data prepared for state timeline**](https://github.com/grafana/grafana/pull/39438/files#diff-07b18f5dca6538123283ce0140166f3dbf0a6a7c06b850b05203dc0c7c5b41ceR52). But this condition was checked against data frames array and fields array length - not taking into the account that data frame may contain multiple fields in it.

Ref https://github.com/grafana/grafana/pull/39438
Ref https://github.com/grafana/grafana/pull/40100

Use the attached dashboard[1] for testing - the 2nd panel does not work on `main` branch.

<details>
<summary>[1] dashboard.json</summary>

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 2,
  "id": 632,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "datasource",
        "uid": "grafana"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "fillOpacity": 80,
            "lineWidth": 0
          },
          "mappings": [
            {
              "options": {
                "CRITICAL": {
                  "color": "red",
                  "index": 3
                },
                "HIGH": {
                  "color": "orange",
                  "index": 2
                },
                "LOW": {
                  "color": "blue",
                  "index": 0
                },
                "NORMAL": {
                  "color": "green",
                  "index": 1
                }
              },
              "type": "value"
            }
          ],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 6,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 65,
      "options": {
        "alignValue": "center",
        "colWidth": 0.9,
        "legend": {
          "calcs": [],
          "displayMode": "table",
          "placement": "bottom"
        },
        "mergeValues": true,
        "mode": "changes",
        "rowHeight": 0.97,
        "showValue": "never",
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "grafana"
          },
          "refId": "s1"
        },
        {
          "datasource": {
            "type": "datasource",
            "uid": "grafana"
          },
          "hide": false,
          "refId": "s2"
        }
      ],
      "title": "State timeline strings",
      "transformations": [],
      "type": "state-timeline"
    },
    {
      "datasource": {
        "type": "testdata",
        "uid": "42joY8pnz"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "fillOpacity": 80,
            "lineWidth": 0
          },
          "mappings": [
            {
              "options": {
                "CRITICAL": {
                  "color": "red",
                  "index": 3
                },
                "HIGH": {
                  "color": "orange",
                  "index": 2
                },
                "LOW": {
                  "color": "blue",
                  "index": 0
                },
                "NORMAL": {
                  "color": "green",
                  "index": 1
                }
              },
              "type": "value"
            }
          ],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 6,
        "w": 24,
        "x": 0,
        "y": 6
      },
      "id": 66,
      "options": {
        "alignValue": "center",
        "colWidth": 0.9,
        "legend": {
          "calcs": [],
          "displayMode": "table",
          "placement": "bottom"
        },
        "mergeValues": true,
        "mode": "changes",
        "rowHeight": 0.97,
        "showValue": "never",
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.0-pre",
      "targets": [
        {
          "csvContent": "\"time\",\"customerid\",\"tax\"\n1075161600000,7888,25.84\n1072915200000,4858,4.53\n\n",
          "datasource": {
            "type": "testdata",
            "uid": "42joY8pnz"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "State timeline strings",
      "transformations": [],
      "type": "state-timeline"
    },
    {
      "datasource": {
        "type": "testdata",
        "uid": "42joY8pnz"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "fillOpacity": 80,
            "lineWidth": 0
          },
          "mappings": [
            {
              "options": {
                "CRITICAL": {
                  "color": "red",
                  "index": 3
                },
                "HIGH": {
                  "color": "orange",
                  "index": 2
                },
                "LOW": {
                  "color": "blue",
                  "index": 0
                },
                "NORMAL": {
                  "color": "green",
                  "index": 1
                }
              },
              "type": "value"
            }
          ],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 6,
        "w": 24,
        "x": 0,
        "y": 12
      },
      "id": 67,
      "options": {
        "alignValue": "center",
        "colWidth": 0.9,
        "legend": {
          "calcs": [],
          "displayMode": "table",
          "placement": "bottom"
        },
        "mergeValues": true,
        "mode": "changes",
        "rowHeight": 0.97,
        "showValue": "never",
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.0-pre",
      "targets": [
        {
          "csvContent": "\"time\",\"tax\"\n1075161600000,25.84\n1072915200000,4.53\n\n",
          "datasource": {
            "type": "testdata",
            "uid": "42joY8pnz"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        },
        {
          "csvContent": "\"time\",\"customerid\"\n1075161600000,7888\n1072915200000,4858\n\n",
          "datasource": {
            "type": "testdata",
            "uid": "42joY8pnz"
          },
          "hide": false,
          "refId": "B",
          "scenarioId": "csv_content"
        }
      ],
      "title": "State timeline strings",
      "transformations": [],
      "type": "state-timeline"
    }
  ],
  "refresh": false,
  "schemaVersion": 34,
  "style": "dark",
  "tags": [
    "demo"
  ],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-15m",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ]
  },
  "timezone": "",
  "title": "State timeline tooltip bug #42620",
  "uid": "qD-rVv6Mz",
  "version": 19,
  "weekStart": ""
}
```
<details>

### CODE REVIEW

**Comment by dprokop** [comment]
@zoltanbedi could you please check whether or not this PR brings back the issues that you have fixed in #39438, #40100?

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 1]
```diff
@@ -39,19 +39,27 @@ export const StateTimelinePanel: React.FC<TimelinePanelProps> = ({
   const renderCustomTooltip = useCallback(
     (alignedData: DataFrame, seriesIdx: number | null, datapointIdx: number | null) => {
       const data = frames ?? [];
+      // Count vaue fields in the state-timeline-ready frame
```
Comment (zoltanbedi) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:42]:
```suggestion
      // Count value fields in the state-timeline-ready frame
```

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 2]
```diff
@@ -39,19 +39,27 @@ export const StateTimelinePanel: React.FC<TimelinePanelProps> = ({
...
+      const valueFieldsCount = data.reduce(
+        (acc, frame) => acc + frame.fields.filter((field) => field.type !== FieldType.time).length,
+        0
+      );
+
       // Not caring about multi mode in StateTimeline
       if (seriesIdx === null || datapointIdx === null) {
         return null;
       }
+      // console.log(seriesIdx, datapointIdx);
```
Comment (zoltanbedi) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:52]:
```suggestion
```

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 3]
```diff
@@ -39,19 +39,27 @@ export const StateTimelinePanel: React.FC<TimelinePanelProps> = ({
...
 
       /**
        * There could be a case when the tooltip shows a data from one of a multiple query and the other query finishes first
        * from refreshing. This causes data to be out of sync. alignedData - 1 because Time field doesn't count.
        * Render nothing in this case to prevent error.
        * See https://github.com/grafana/support-escalations/issues/932
        */
+      // console.log('retununi', alignedData.fields.length - 1, data);
```
Comment (zoltanbedi) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:60]:
```suggestion
```

**Review by zoltanbedi — APPROVED** [review]
LGTM, thanks for fixing this @dprokop. I've also tried with my dashboard with the original issue and the fix still works there as well.

**Pull request merged**